### PR TITLE
OCLOMRS-303: Create concept mapping form

### PR DESCRIPTION
**Pull request by Duncanian**
# JIRA TICKET NAME:
[Create concept mapping form](https://issues.openmrs.org/browse/OCLOMRS-303)

# Summary:
Concept mapping is an instructional strategy for individual and group learning that involves the integration of knowledge and creation of meaning from each individual's language or locale that relates to specific concepts. This ticket involves the creation of the mapping forms without its functionalities.

### CODE REVIEW

File: src/components/dictionaryConcepts/components/CreateConceptForm.jsx [hunk 1]
```diff
@@ -174,6 +176,37 @@ const CreateConceptForm = props => (
           </a>
         </div>
       </div>
+
+      {props.isEditConcept && (
```
Comment (dessHub) [inline src/components/dictionaryConcepts/components/CreateConceptForm.jsx]: `props.*` has been used more than once, you can destructure.

File: src/components/dictionaryConcepts/components/ActionButtons.jsx [hunk 2]
```diff
@@ -1,27 +1,36 @@
...
+          </Link>
+          <AddMapping />
+        </React.Fragment>
       )}
       <button
         type="button"
         className="btn btn-sm mb-1 actionButtons"
         id="retireConcept"
         onClick={() => { showDeleteModal(version_url); }}
+        style={{ marginLeft: '12px' }}
```
Comment (muhozi) [inline src/components/dictionaryConcepts/components/ActionButtons.jsx]: @Duncanian Why are you adding styles here?

File: src/components/dictionaryConcepts/components/AddMapping.jsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    });
+  };
+
+  render() {
+    return (
+      <React.Fragment>
+        <button
+          type="button"
+          className="btn btn-sm mb-1 actionButtons"
+          // id="retireConcept"
```
Comment (muhozi) [inline src/components/dictionaryConcepts/components/AddMapping.jsx]: @Duncanian  please remove this comment.

File: src/components/dictionaryConcepts/components/ExternalMapping.jsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      term,
+      source_url,
+      code,
+    } = this.props;
+
+    return (
+      <div>
+        <form>
+          <div>
+            <FormGroup style={{ marginTop: '12px' }}>
```
Comment (muhozi) [inline src/components/dictionaryConcepts/components/ExternalMapping.jsx]: @Duncanian Why are you putting styles here?

File: src/components/dictionaryConcepts/components/ExternalMapping.jsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                id="map_type"
+                required
+              >
+                <option>Same as</option>
+                <option>Narrower than</option>
+              </select>
+            </FormGroup>
+          </div>
+          <div>
+            <FormGroup style={{ marginTop: '12px' }}>
```
Comment (muhozi) [inline src/components/dictionaryConcepts/components/ExternalMapping.jsx]: Same, why are you putting styles here?

File: src/components/dictionaryConcepts/components/ActionButtons.jsx (hunk 2 above)
Comment (Duncanian) [reply src/components/dictionaryConcepts/components/ActionButtons.jsx]: @muhozi , let me refactor.

File: .eslintrc.json [hunk 6]
```diff
@@ -18,7 +18,8 @@
     "jsx-a11y/no-static-element-interactions": 0,
     "compilerOptions": { "target": "ES6", "module": "commonjs" },
     "max-len": ["error", { "code": 100, "ignoreStrings": true, "ignoreTemplateLiterals": true }],
-    "import/no-extraneous-dependencies": ["error", {"devDependencies": true}]
+    "import/no-extraneous-dependencies": ["error", {"devDependencies": true}],
+    "react/prefer-stateless-function": 0
```
Comment (muhozi) [inline .eslintrc.json]: Why are you adding this rule? Was it related to this ticket?

Comment (Duncanian) [reply .eslintrc.json]: @muhozi , yeah, it was related to the types of components I should use.

File: src/components/dictionaryConcepts/components/AddMapping.jsx (hunk 3 above)
Comment (Duncanian) [reply src/components/dictionaryConcepts/components/AddMapping.jsx]: @muhozi , let me refactor.

File: src/components/dictionaryConcepts/components/ExternalMapping.jsx (hunk 4 above)
Comment (Duncanian) [reply src/components/dictionaryConcepts/components/ExternalMapping.jsx]: @muhozi , let me refactor.

File: src/components/dictionaryConcepts/components/ExternalMapping.jsx (hunk 5 above)
Comment (Duncanian) [reply src/components/dictionaryConcepts/components/ExternalMapping.jsx]: @muhozi , let me refactor.

File: src/tests/dictionaryConcepts/components/AddMapping.test.jsx [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+describe('render MappingModal', () => {
+  beforeEach(() => {
+    wrapper = shallow(<AddMapping />);
+  });
+  it('should render without breaking', () => {
+    expect(wrapper.length).toEqual(1);
+  });
+
+  it('should handle type change', () => {
+    wrapper.find('.actionButtons').simulate('click');
```
[the remark this answers is not in the document]
Comment (Duncanian) [reply src/tests/dictionaryConcepts/components/AddMapping.test.jsx]: @Daramola98 , Am making the changes.

File: src/tests/dictionaryConcepts/components/ExternalMapping.test.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  it('should handle input change', () => {
+    // const spy = jest.spyOn(wrapper.instance(), 'handleType');
+    const event = {
+      target: {
+        name: 'type',
+        value: 'test',
+      },
+    };
+    wrapper.find('#map_type').simulate('change', event);
+    // expect(spy).toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (Duncanian) [reply src/tests/dictionaryConcepts/components/ExternalMapping.test.js]: @Daramola98 , Am making the changes.

File: src/tests/dictionaryConcepts/components/InternalMapping.test.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+      handleChange: jest.fn(),
+    };
+    wrapper = shallow(<InternalMapping {...props} />);
+  });
+  it('should render without breaking', () => {
+    expect(wrapper.length).toEqual(1);
+  });
+
+  it('should handle input change', () => {
+    // const spy = jest.spyOn(wrapper.instance(), 'handleType');
```
[the remark this answers is not in the document]
Comment (Duncanian) [reply src/tests/dictionaryConcepts/components/InternalMapping.test.js]: @Daramola98 , Am making the changes.

File: src/tests/dictionaryConcepts/components/MappingModal.test.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  });
+
+  it('should handle change of type', () => {
+    const event = {
+      target: {
+        name: 'type',
+        value: 'test',
+      },
+    };
+    wrapper.find('#type').simulate('change', event);
```
[the remark this answers is not in the document]
Comment (Duncanian) [reply src/tests/dictionaryConcepts/components/MappingModal.test.js]: @Daramola98 , Am making the changes.

**Pull request merged**